### ISSUE TITLE
View is built but not showing as such in view_build_status system table

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylladb#6094]
Seen in https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-release/442/testReport/materialized_views_test/TestMaterializedViews/mv_populating_from_existing_data_during_node_restart_test/
Scylla version c639a5ec6f081c857b4f62ab3501b74347ff52ee
scylladb/scylla-dtest@6caf0ec86e08b90b9d137e8a91c90dde08bcbf82

I'm not 100% certain this is a bug in scylla and not in the test...
Anyhow, the test failed as follows:
```
Error Message
View ks.tm_table_mv_1 not built
Stacktrace
Traceback (most recent call last):
  File "/usr/lib64/python3.7/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/usr/lib64/python3.7/unittest/case.py", line 645, in run
    testMethod()
  File "/jenkins/workspace/scylla-master/dtest-release/scylla-dtest/materialized_views_test.py", line 616, in mv_populating_from_existing_data_during_node_restart_test
    self._mv_populating_from_existing_data_during_changes_test('restart node', nodes=4, rf=3, mvs=10, prefill=40000)
  File "/jenkins/workspace/scylla-master/dtest-release/scylla-dtest/materialized_views_test.py", line 671, in _mv_populating_from_existing_data_during_changes_test
    wait_for_view(cluster=self.cluster, session=session, ks=tm.keyspace, view=mv_name)
  File "/jenkins/workspace/scylla-master/dtest-release/scylla-dtest/scylla_tools.py", line 933, in wait_for_view
    raise Exception(error_msg)
Exception: View ks.tm_table_mv_1 not built
```

`wait_for_view` looks for all live nodes' uuid in `"SELECT {} FROM system_distributed.view_build_status"`

https://github.com/scylladb/scylla-dtest/blob/6caf0ec86e08b90b9d137e8a91c90dde08bcbf82/scylla_tools.py#L904-L923

Despite it failing, we see that the view was built on all nodes' logs:
```
$ grep 'Finished building view ks.tm_table_mv_1' logs-release.2/1585468768826_materialized_views_test.TestMaterializedViews.mv_populating_from_existing_data_during_node_restart_test/node?.log

logs-release.2/1585468768826_materialized_views_test.TestMaterializedViews.mv_populating_from_existing_data_during_node_restart_test/node1.log:INFO  2020-03-29 07:43:57,715 [shard 0] view - Finished building view ks.tm_table_mv_1

logs-release.2/1585468768826_materialized_views_test.TestMaterializedViews.mv_populating_from_existing_data_during_node_restart_test/node2.log:INFO  2020-03-29 07:44:32,308 [shard 0] view - Finished building view ks.tm_table_mv_1

logs-release.2/1585468768826_materialized_views_test.TestMaterializedViews.mv_populating_from_existing_data_during_node_restart_test/node3.log:INFO  2020-03-29 07:43:55,481 [shard 0] view - Finished building view ks.tm_table_mv_1
logs-release.2/1585468768826_materialized_views_test.TestMaterializedViews.mv_populating_from_existing_data_during_node_restart_test/node3.log:INFO  2020-03-29 07:43:56,666 [shard 0] view - Finished building view ks.tm_table_mv_1

logs-release.2/1585468768826_materialized_views_test.TestMaterializedViews.mv_populating_from_existing_data_during_node_restart_test/node4.log:INFO  2020-03-29 07:43:55,029 [shard 0] view - Finished building view ks.tm_table_mv_1
logs-release.2/1585468768826_materialized_views_test.TestMaterializedViews.mv_populating_from_existing_data_during_node_restart_test/node4.log:INFO  2020-03-29 07:43:56,122 [shard 0] view - Finished building view ks.tm_table_mv_1
```

And the [dtest.log](https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-release/442/artifact/logs-release.2/dtest.log) debug log shows:
```
2020-03-29 07:44:43,154 17807   dtest                          DEBUG    | materialized_views_test.py:TestMaterializedViews.mv_populating_from_existing_data_during_node_restart_test - Waiting for view ks.tm_table_mv_1 to finish building...
2020-03-29 07:59:28,811 17807   dtest                          DEBUG    | materialized_views_test.py:TestMaterializedViews.mv_populating_from_existing_data_during_node_restart_test - Test failed with errors: [(<materialized_views_test.TestMaterializedViews testMethod=mv_populating_from_existing_data_during_node_restart_test>, (<class 'Exception'>, Exception('View ks.tm_table_mv_1 not built'), <traceback object at 0x7f01db3a80f0>))]
```

**Comment by bhalevy** [comment]
@eliransin / @psarna please take a look

**Comment by bhalevy** [comment]
Why do we see 2 reports for `Finished building view ks.tm_table_mv_1` on nodes 3 and 4?

**Comment by psarna** [comment]
Double `Finished building view ks.tm_table_mv_1` is caused by a schema change occurring in the middle of view building, here's the whole context:
```
INFO  2020-03-29 07:43:55,481 [shard 0] view - Finished building view ks.tm_table_mv_1
(...)
INFO  2020-03-29 07:43:55,598 [shard 0] schema_tables - Altering ks.tm_table id=eb583fc0-7190-11ea-b53d-000000000000 version=c997ee61-a101-3778-9e31-29dcc7c7c37f
INFO  2020-03-29 07:43:55,599 [shard 0] schema_tables - Dropping ks.tm_table_mv_1 id=082e2560-7191-11ea-b9d1-000000000000 version=57a30edc-29c4-3c9a-8934-12054e5a4612
INFO  2020-03-29 07:43:55,599 [shard 0] schema_tables - Creating ks.tm_table_mv_1 id=09df7ee0-7191-11ea-b53d-000000000000 version=d7eaa2e4-8af7-335a-999a-bb80a09b64cf
(...)
INFO  2020-03-29 07:43:55,990 [shard 0] view - Stopping to build view ks.tm_table_mv_1
INFO  2020-03-29 07:43:55,990 [shard 0] query_processor - Column definitions for ks.tm_table changed, invalidating related prepared statements
INFO  2020-03-29 07:43:55,991 [shard 0] view - Building view ks.tm_table_mv_1, starting at token minimum token
INFO  2020-03-29 07:43:56,064 [shard 0] database - Schema version changed to cfe21a52-86a6-33af-998a-e3af86963abf
INFO  2020-03-29 07:43:56,666 [shard 0] view - Finished building view ks.tm_table_mv_1
```
I can't reproduce this failure, but, first of all, this is one of the test which run without schema agreement:
```python
         debug('Disabling schema agreement')
         session.cluster.max_schema_agreement_wait = 0
```
... that's because otherwise, while we wait for schema agreement, the views could have already finished building and this test would lose most of its sense, so not waiting for schema agreement has merit. At the same time, it looks like it may also be the cause of flakiness. I'll try to run this test case a couple of more thousand times and see if it reproduces, because otherwise I'm not sure how to remedy this. Perhaps due to not waiting for schema to agree, some uuids can get stale and the test is trying to assert with an outdated value?

**Comment by nyh** [comment]
I don't understand how the schema agreement hypothesis explains this problem. As `view_built_status_query()` in `scylla_tools.py` shows, the list of fully built views is kept by view **name**, not UUID or anything of this sort. So if the log reports that we finished building a table, how does the schema changing explain why we don't see it in the table of finished builds?

**Comment by psarna** [comment]
@nyh right, uuids are used only for `host_id`, but I think we can still expect some flakiness from racing with schema agreements, e.g.:
1. We want to be able to assume that all views are built for our test case, so we call `wait_for_view`
2. A wild schema change appears since we didn't wait for a schema agreement, and it causes our view to be recreated from scratch
3. The assumption is no longer true - a view that was already marked as `built` is now `in progress`

That said, it's probably not the issue in this test case, because waiting for view building to finish is what fails - or, rather, times out.

I wonder if it's possible that we finished view building, but just failed to keep up-to-date information in the view status table. Note that in order to classify the view as built, we wait for **all** `<node_id, status>` pairs for all nodes, so perhaps we only partially failed to save the information. This test involves restarting a node, so maybe it was restarted precisely when we wanted to mark the view as finished, and due to some kind of race condition, we lost the information? I'm just hand-waving here, I wish I could reproduce this failure locally... neither `dev` nor `release` triggered even once.

**Comment by nyh** [comment]
@psarna yes, my point was that regardless of whether or not schema changes can cause problems in view building (and I have to be honest, I'm not sure), it can't be the issue we're looking for here so let's not discuss them in the context of this issue (you can open a new issue if you think it is a problem).

Your latest guess at the end of your last comment seems more likely. In the past, we had a problem in the wait_for_view() function when one of the nodes went down, because it insisted to wait for that node as well to finish its building, which would of course never happen (because the node is down!). So the current implementation of wait_for_view(), if I remember correctly, only wait for nodes which are up to finish building. But maybe there is a race here - maybe when wait_for_view() decided on which node to look at, some node A was alive, but only then, this node A went down and wait_for_view() couldn't see that it finished the build. I thought I considered these issues and fixed them in the past (so as soon as a node goes down we stop waiting for it to build) but maybe there's an edge condition I didn't consider.

**Comment by psarna** [comment]
Right. This particular case is a restart, so the node should probably eventually come back to life, but then, it might also simply be a timing issue - i.e. a node failed to stop&start in the arbitrary 60 seconds window, which is the default timeout for wait_for_view. Doesn't seem entirely unlikely, given that booting itself can be a lengthy process.

**Comment by psarna** [comment]
Oh wait, it's actually 5 minutes of waiting (60x5mins). That should be enough to acknowledge everything... I'll check the logs again to look for inspiration.

**Comment by bhalevy** [comment]
> Oh wait, it's actually 5 minutes of waiting (60x5mins). That should be enough to acknowledge everything... I'll check the logs again to look for inspiration.

@psarna I speculate that the schema change might have something to do with this.
`view_builder::on_drop_view` is runs a background task that eventually calls `_sys_dist_ks.remove_view`

and in node4 logs we see:
```
INFO  2020-03-29 07:43:55,503 [shard 0] schema_tables - Dropping ks.tm_table_mv_1 id=082e2560-7191-11ea-b9d1-000000000000 version=57a30edc-29c4-3c9a-8934-12054e5a4612
INFO  2020-03-29 07:43:55,503 [shard 0] schema_tables - Creating ks.tm_table_mv_1 id=09df7ee0-7191-11ea-b53d-000000000000 version=d7eaa2e4-8af7-335a-999a-bb80a09b64cf

INFO  2020-03-29 07:43:55,762 [shard 0] view - Stopping to build view ks.tm_table_mv_1

^^^

INFO  2020-03-29 07:43:55,763 [shard 0] query_processor - Column definitions for ks.tm_table changed, invalidating related prepared statements
INFO  2020-03-29 07:43:55,763 [shard 1] query_processor - Column definitions for ks.tm_table changed, invalidating related prepared statements
INFO  2020-03-29 07:43:55,765 [shard 0] view - Building view ks.tm_table_mv_1, starting at token minimum token
INFO  2020-03-29 07:43:55,827 [shard 0] database - Schema version changed to cfe21a52-86a6-33af-998a-e3af86963abf

INFO  2020-03-29 07:43:56,074 [shard 0] migration_manager - Create new view: org.apache.cassandra.config.CFMetaData@0x600003a00000[cfId=0b3990a0-7191-11ea-9091-000000000000,ksName=ks,cfName=tm_table_mv_2,cfType=Standard,comparator=org.apache.cassandra.db.marshal.CompositeType(org.apache.cassandra.db.marshal.Int32Type,org.apache.cassandra.db.marshal.UTF8Type),comment=,readRepairChance=0,dcLocalReadRepairChance=0.1,gcGraceSeconds=864000,keyValidator=org.apache.cassandra.db.marshal.IntegerType,minCompactionThreshold=4,maxCompactionThreshold=32,columnMetadata=[ColumnDefinition{name=id, type=org.apache.cassandra.db.marshal.IntegerType, kind=PARTITION_KEY, componentIndex=0, droppedAt=-9223372036854775808}, ColumnDefinition{name=clmn_int2, type=org.apache.cassandra.db.marshal.Int32Type, kind=CLUSTERING_COLUMN, componentIndex=0, droppedAt=-9223372036854775808}],compactionStrategyClass=class org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy,compactionStrategyOptions={enabled=true},compressionParameters={sstable_compression=org.apache.cassandra.io.compress.LZ4Compressor},bloomFilterFpChance=0.01,memtableFlushPeriod=0,caching={"keys":"ALL","rows_per_partition":"ALL"},cdc={},defaultTimeToLive=0,minIndexInterval=128,maxIndexInterval=2048,speculativeRetry=99.0PERCENTILE,triggers=[],isDense=false,version=0b3990a1-7191-11ea-9091-000000000000,droppedColumns={},collections={},indices={}, viewInfo=ViewInfo{baseTableId=eb583fc0-7190-11ea-b53d-000000000000, baseTableName=tm_table, includeAllColumns=false, whereClause=id IS NOT null AND clmn_int2 IS NOT null}]

INFO  2020-03-29 07:43:56,122 [shard 0] view - Finished building view ks.tm_table_mv_1
```

`system_distributed_keyspace::remove_view` deletes the row for for the view (by name) from view_build_status:
```
            format("DELETE FROM {}.{} WHERE keyspace_name = ? AND view_name = ?", NAME, VIEW_BUILD_STATUS),
```

and `system_distributed_keyspace::finish_view_build` tries to UPDATE it (not INSERT):
```
                format("UPDATE {}.{} SET status = ? WHERE keyspace_name = ? AND view_name = ? AND host_id = ?", NAME, VIEW_BUILD_STATUS),
```

So if it's deleted before updated we're screwed, right?

**Comment by bhalevy** [comment]
Should we keep the version uuid in view_build_status to uniquely identify the view?

**Comment by psarna** [comment]
@bhalevy not so, inserting via `UPDATE` is also a valid way to create a Scylla row - the difference is that in case of `UPDATE WHERE`, only the columns keep the row alive, there's no row marker.

To demonstrate the difference, if you have this table:
```cql
CREATE TABLE t(id int primary key, v int);
```
and insert two rows like that:
```cql
INSERT INTO t (id, v) VALUES (1, 1);
UPDATE t SET v = 2 WHERE id = 2;
```
and then remove v from both:
```cql
DELETE v FROM t WHERE id = 1;
DELETE v FROM t WHERE id = 2;
```
... then the first row will still be alive, while the second will disappear:
```cql
SELECT * FROM t;
 id | v
----+------
  1 | null
(1 rows)
```

**Comment by nyh** [comment]
But @psarna maybe the issue isn't insert vs update, but rather the timing of these updates or deletes: Maybe it's possible (I don't know...) that a millisecond after the view finished building and inserted (updated) the build-finished row, another piece of code that noticed that the old schema was deleted went and **deleted** this row. If this is what happens (I didn't check the code...), I think maybe we make a mistake by deleting the build-finished row when a schema is deleted. This row should be deleted only when the last schema with the same **name** is deleted, or equivalently when we really **drop** the table - just modifying the table's schema shouldn't cause this row to be removed.

**Comment by psarna** [comment]
Indeed, I'm fairly certain that the issue is not insert vs update, the example above was just trying to show that it should *not* be problematic :)

We already had a similar issue related to view building bookkeeping: https://github.com/scylladb/scylla/commit/fb37394995da8051057fbd4c67ab9df05052d189, where the recreation of a view could result in it disappearing from bookkeeping. Perhaps there's still a race in there somewhere... the problem is that both adding new bookkeeping information and removing bookkeeping information are serialized by a semaphore, so theoretically they should never be executed in a wrong order (e.g. insert new view information with an identical name first, and then remove the old view information, resulting in removal of the new view, because they have the same name).

**But**, there's something I don't understand here - there are two paths of registering new views in the builder. One, guarded by a semaphore, is `view_builder::on_create_view` callback, and the other, **not guarded by any semaphore**, is in `calculate_shard_build_step`:
```cpp
(...)
    auto all_views = _db.get_views();
    auto is_new = [&] (const view_ptr& v) {
        return base_table_exists(v) && loaded_views.find(v->id()) == loaded_views.end()
                && built_views.find(v->id()) == built_views.end();
    };
    for (auto&& view : all_views | boost::adaptors::filtered(is_new)) {
        bookkeeping_ops->push_back(add_new_view(view, get_or_create_build_step(view->view_info()->base_id())));
    }
(...)
```
- and that suggests that the view can get added outside of the semaphore.

Hm, maybe this kind of a race is possible (judged by how node3 logs look like):
1. We see `Stopping to build view ks.tm_table_mv_1` in the logs, which starts a background operation:
    a. wait on the semaphore
    b. drop the view information (based on its name)
2. We see `Building view ks.tm_table_mv_1, starting at token minimum token`, which means that the view build started - but let's assume it wasn't started via `view_builder::on_create_view`, which is serialized by a semaphore, but directly via `calculate_shard_build_step`, which doesn't use semaphores and can be executed right away
3. View building finishes
4. Only now, the continuation from 1b wakes up and deletes the old view information, which happens to have the same name, so ends up in deleting the *new* view information instead

Now, the state is like that: the view is build, but bookkeeping does not store any information about it, because it was deleted by background continuation from 1b.
Reasons why this reasoning is far-fetched:
1. I don't exactly see why we have two ways of registering new views... wouldn't `on_create_view` be unconditionally called each time a view is created? Or does it not happen in certain cases, e.g. at startup?
2. If a view is recreated, then **both** `view_builder::on_drop_view` and `view_builder::on_create_view` should be eventually called, and their order is serialized with a semaphore - so even if we temporarily have incorrect bookkeeping after dropping the view ends, the worst that should happen is that shortly after, `view_builder::on_create_view` tries to build the view again - which is redundant, but not incorrect.

Opinions?

**Comment by bhalevy** [comment]
@nyh can you please state your opinion on the above suggestion?
I'm don't feel familiar enough with the MV implementation to have a valuable opinion about it.

**Comment by nyh** [comment]
@psarna, I'm not fluent in these code paths, Duarte wrote them initially and you improved on them, so I can't say I understand exactly where the bug is, but here is my own guess (which is somewhat different than your guess above):

Imagine that the view builder is happily building a view, calling `calculate_shard_build_step()`  in a loop. Now, in parallel, somebody removes and re-adds this view. The removal happens first, and at that point the `calculate_shard_build_step()` which was already running wakes up and notices the view has disappeared and it can't calculate the next step. In that case, its code runs `remove_built_view()` to remove this build progress. But this remove_built_view is a future, it can be delayed indefinitely.
Now, before the remove_built_view action actually happens, we could have a lot of other stuff happen: Consider the view replacement continues - as you said on_drop_view() is called, and only then the view is re-added, on_create_view() called and the entire view is rebuilt and is marked built. ONLY THEN, finally (!), the remove_built_view() which we started a while back, gets run and removes the information on the just-built view.

I think the chances of this happening is very low (the view build will have to be super-small and quick!) but not impossibe.

If my guess is correct, perhaps one way to fix this is to add a semaphore serializing just the `remove_built_view()` and `mark_view_as_built()` calls with a semaphore (it could be a separate semaphore per view name, but it's easier to just have one semaphore - it's not like we have a lot of these operations). I.e., if we had already decided to mark a view-build removed, let's not allow the same view  to be completely built and marked built before we really finished marking it removed.

**Comment by psarna** [comment]
@nyh Makes sense, especially that you mentioned one thing that is reflected in the logs of this issue - the view building was indeed super quick.
I think I'll try a shot in the dark - I'll prepare a patch that serializes more bookkeeping operations under the semaphore (the ones that you mentioned and perhaps some other ones if I find them) - bookkeeping the events related to build creation/deletion are considered very rare, as you correctly stated, which means they're not on a hot path, so we can afford to be extra careful.

**Comment by psarna** [comment]
Hm, it's actually more convoluted. `calculate_shard_build_step` is only called at startup, not in a loop. That said, it's still outside of any semaphore, so it still seems possible that it can race with `on_drop_view`. I think that it's actually enough to just wrap this entire initial step with a semaphore as well, since it's capable of removing view info.

**Comment by nyh** [comment]
@psarna I thought it's not called in a "loop" simplistically, but still called over and over to calculate the next step. Otherwise, who calculates the next build step, after the first one finished?

You are right that if on_create_view() cannot be called before on_drop_view() ends (that's your previous patch) and you change it so on_drop_view() cannot start before calculate_shared_build_step() ends, then this might be indeed enough serialization.

(something to check though, I don't remember - do we allow view-building multiple different tables in parallel? if we do, we need to watch out not to serialize them).

**Comment by psarna** [comment]
@nyh the names may be misleading (they also mislead me until today :)) - `calculate_shard_build_step` is actually `since it looks like the node (re)started, let's calculate from which token we need to restart the view building process on each shard`. From there,  each next step is kept as a next token, since they're ordered, so we don't need to recalculate anything. And as for view-building multiple different tables in parallel - yes and no. We do of course allow interleaving the build steps of different tables, but the builder only executes a single step per shard at a time - which was mentioned in the original commit message as a mechanism of limiting the impact of view building for a cluster.

**Comment by slivne** [comment]
@psarna please check if the same issue occurred in, and if your patch was merged 
https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-release/451/testReport/materialized_views_test/TestMaterializedViews/mv_populating_from_existing_data_during_node_restart_test/

**Comment by psarna** [comment]
@slivne it was run with the patch merged and the beginning of this issue is the same - schema update arrives in the middle of building view4, so the view building is restarted, but I think that the root cause is different... Since this test involves a node restart, here's what happens:
1. View building finishes on node1
2. In the middle of updating `system_distributed.view_build_status` table, node4 restarts:
```
NFO  2020-04-09 03:12:12,993 [shard 0] view - Finished building view ks.tm_table_mv_7
INFO  2020-04-09 03:12:12,993 [shard 0] view - Finished building view ks.tm_table_mv_6
INFO  2020-04-09 03:12:12,993 [shard 0] view - Finished building view ks.tm_table_mv_5
INFO  2020-04-09 03:12:12,993 [shard 0] view - Finished building view ks.tm_table_mv_4
INFO  2020-04-09 03:12:13,140 [shard 0] view - Finished building view ks.tm_table_mv_8
INFO  2020-04-09 03:12:13,220 [shard 0] view - Finished building view ks.tm_table_mv_9
INFO  2020-04-09 03:12:13,924 [shard 0] rpc - client 127.0.50.2:7000: fail to connect: Connection refused
INFO  2020-04-09 03:12:14,927 [shard 0] rpc - client 127.0.50.2:7000: fail to connect: Connection refused
```
3. The information is not propagated to node4 and the error is ignored, since we explicitly do so (to preserve availability I presume):
```cpp
return _qp.execute_internal(
        format("UPDATE {}.{} SET status = ? WHERE keyspace_name = ? AND view_name = ? AND host_id = ?", NAME, VIEW_BUILD_STATUS),
        db::consistency_level::ONE,
        internal_distributed_timeout_config,
        { "SUCCESS", std::move(ks_name), std::move(view_name), std::move(host_id) },
        false).discard_result(); /// <--- we discard the result
```
4. View status kept in `system_distributed.view_build_status` is both **written to** and **read from** with `consistency_level::ONE` - perhaps that's the source of `wait_for_view` thinking that view building hasn't finished?

To sum up, it still looks like messy bookkeeping, since the view is actually built - it's just somehow not acknowledged properly by dtest's `wait_for_view`... perhaps we should query with higher consistency level?

**Comment by bhalevy** [comment]
> perhaps we should query with higher consistency level?

+1

**Comment by DoronArazii** [comment]
@asias can we close it?

**Comment by bhalevy** [comment]
We haven't seen this for a very long while.
Closing...
